lib/ansible/modules/crypto/openssl_privatekey_info.py

```diff
@@ -0,0 +1,351 @@
+                passphrase=to_bytes(self.passphrase) if self.passphrase is not None else self.passphrase,
+                backend=self.backend
+            )
+            result['can_parse_key'] = True
+        except crypto_utils.OpenSSLObjectError as exc:
+            self.module.fail_json(msg=to_native(exc), **result)
+
+        result['public_key'] = self._get_public_key(binary=False)
+        result['public_key_fingerprints'] = crypto_utils.get_fingerprint_of_bytes(self._get_public_key(binary=True))
+
+        key_type, key_public_data, key_private_data = self._get_key_info()
+        result['type'] = key_type
+        result['public_data'] = key_public_data
+        if self.return_private_key_data:
+            result['private_data'] = key_private_data
+
+        return result
+
+
+class PrivateKeyInfoCryptography(PrivateKeyInfo):
+    """Validate the supplied private key, using the cryptography backend"""
+    def __init__(self, module):
+        super(PrivateKeyInfoCryptography, self).__init__(module, 'cryptography')
+
+    def _get_public_key(self, binary):
+        return self.key.public_key().public_bytes(
+            serialization.Encoding.DER if binary else serialization.Encoding.PEM,
+            serialization.PublicFormat.SubjectPublicKeyInfo
+        )
+
+    def _get_key_info(self):
+        return _get_cryptography_key_info(self.key)
+
+
+class PrivateKeyInfoPyOpenSSL(PrivateKeyInfo):
+    """validate the supplied private key."""
+
+    def __init__(self, module):
+        super(PrivateKeyInfoPyOpenSSL, self).__init__(module, 'pyopenssl')
+
+    def _get_public_key(self, binary):
+        try:
+            return crypto.dump_publickey(
+                crypto.FILETYPE_ASN1 if binary else crypto.FILETYPE_PEM,
+                self.key
+            )
+        except AttributeError:
+            self.module.warn('Your pyOpenSSL version does not support dumping public keys. '
+                             'Please upgrade to version 16.0 or newer, or use the cryptography backend.')
+
+    def _get_key_info(self):
+        if PYOPENSSL_VERSION >= LooseVersion('16.1.0') and CRYPTOGRAPHY_FOUND:
+            return _get_cryptography_key_info(self.key.to_cryptography_key())
+        key_public_data = dict()
+        key_private_data = dict()
+        if self.key.type() == crypto.TYPE_RSA:
+            key_type = 'RSA'
+            key_public_data['size'] = self.key.bits()
+        elif self.key.type() == crypto.TYPE_DSA:
+            key_type = 'DSA'
+            key_public_data['size'] = self.key.bits()
+        else:
+            key_type = 'unknown ({0})'.format(self.key.type())
+        return key_type, key_public_data, key_private_data
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            path=dict(type='path', required=True),
+            passphrase=dict(type='str', no_log=True),
+            return_private_key_data=dict(type='bool', default=True),
+            valid_at=dict(type='dict'),
```

Comment:
huh?

Reply:
Guess where I copied the module from... :D